sympy/core/tests/test_parameters.py

```diff
@@ -88,3 +88,27 @@ def test_nested():
         expr = (x + x) + (y + y)
         assert expr.args == ((x + x), (y + y))
         assert expr.args[0].args == (x, x)
+
+def test_reentrantcy():
+    with evaluate(False):
+        expr = x + x
+        assert expr.args == (x, x)
+        with evaluate(True):
+            expr = x + x
+            assert expr.args == (2, x)
+        expr = x + x
+        assert expr.args == (x, x)
+
+def test_reusability():
+    f = evaluate(False)
+
+    with f:
```

Comment:
Is this why `evaluate` became a class?

I don't really mind but I don't think it is really necessary to support it being used in this way.

Reply:
Yes. I had some code that I was going to use in the other PR that made use of this. That code ended up being scrapped, but this still seemed useful to have anyway.